lntemp/node/watcher.go

```diff
@@ -0,0 +1,724 @@
+package node
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightningnetwork/lnd/lnrpc"
+	"github.com/lightningnetwork/lnd/lntemp/rpc"
+	"github.com/lightningnetwork/lnd/lntest"
+	"github.com/lightningnetwork/lnd/lntest/wait"
+)
+
+type chanWatchType uint8
+
+const (
+	// watchOpenChannel specifies that this is a request to watch an open
+	// channel event.
+	watchOpenChannel chanWatchType = iota
+
+	// watchCloseChannel specifies that this is a request to watch a close
+	// channel event.
+	watchCloseChannel
+
+	// watchPolicyUpdate specifies that this is a request to watch a policy
+	// update event.
+	watchPolicyUpdate
+
+	// TODO(yy): remove once temp tests is finished.
+	DefaultTimeout = lntest.DefaultTimeout
+)
+
+// closeChanWatchRequest is a request to the lightningNetworkWatcher to be
+// notified once it's detected within the test Lightning Network, that a
+// channel has either been added or closed.
+type chanWatchRequest struct {
+	chanPoint wire.OutPoint
+
+	chanWatchType chanWatchType
+
+	eventChan chan struct{}
+
+	advertisingNode    string
+	policy             *lnrpc.RoutingPolicy
+	includeUnannounced bool
+}
+
+// nodeWatcher is a topology watcher for a HarnessNode. It keeps track of all
+// the topology updates seen in a given node, including NodeUpdate,
+// ChannelEdgeUpdate, and ClosedChannelUpdate.
+type nodeWatcher struct {
+	// rpc is the RPC clients used for the current node.
+	rpc *rpc.HarnessRPC
+
+	// state is the node's current state.
+	state *State
+
+	// chanWatchRequests receives a request for watching a particular event
+	// for a given channel.
+	chanWatchRequests chan *chanWatchRequest
+
+	// For each outpoint, we'll track an integer which denotes the number
+	// of edges seen for that channel within the network. When this number
+	// reaches 2, then it means that both edge advertisements has
+	// propagated through the network.
+	// openChanWatchers  map[wire.OutPoint][]chan struct{}
+	openChanWatchers sync.Map
```

Comment:
In the `state.go` we use pointers to these maps. Since maps shouldn't be copied (docs state: "A Map must not be copied after first use."), we should probably use pointers everywhere?